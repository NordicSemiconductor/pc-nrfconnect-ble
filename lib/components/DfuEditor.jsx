/* Copyright (c) 2016, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *   1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form, except as embedded into a Nordic
 *   Semiconductor ASA integrated circuit in a product or a software update for
 *   such product, must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 *
 *   3. Neither the name of Nordic Semiconductor ASA nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 *   4. This software, with or without modification, must only be used with a
 *   Nordic Semiconductor ASA integrated circuit.
 *
 *   5. Any software provided in binary form under this license must not be
 *   reverse engineered, decompiled, modified and/or disassembled.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React, { PropTypes } from 'react';
import { FormGroup, ControlLabel, ButtonToolbar } from 'react-bootstrap';

import FileInput from './input/FileInput';
import ProgressBarInput from './input/ProgressBarInput';
import ReadOnlyField from './input/ReadOnlyField';
import DfuButton from './DfuButton';
import DfuThroughputGraph from './DfuThroughputGraph';

/*
 * Converts an object like:
 *
 * { "application": { "dat_file": "file.dat", "bin_file": "file.bin" }, ... }
 *
 * to:
 *
 * application:
 *   dat_file: file.dat
 *   bin_file: file.bin
 * ...
 */
function createPackageInfoString(packageInfo) {
    return Object.keys(packageInfo).reduce((prevType, type) => {
        const valueString = Object.keys(packageInfo[type]).reduce((prevValue, value) => {
            return prevValue + '\n  ' + value + ': ' + packageInfo[type][value];
        }, '');
        return prevType + type + ':' + valueString + '\n';
    }, '');
}

class DfuEditor extends React.PureComponent {
    renderPackageInfo() {
        return (
            <ReadOnlyField
                label='Package info'
                value={createPackageInfoString(this.props.packageInfo)}
            />
        );
    }

    renderProgress() {
        const percentCompleted = this.getPercentCompleted();
        return (
            <ProgressBarInput
                label='Progress'
                status={this.getStatus()}
                now={percentCompleted}
                progressLabel={`${percentCompleted}%`}
            />
        );
    }

    renderGraph() {
        const throughput = this.props.throughput;
        if (throughput && throughput.kbpsPoints.length > 0) {
            return (
                <FormGroup>
                    <ControlLabel className="col-md-3 text-right">Throughput (kB/s)</ControlLabel>
                    <div className="col-md-9">
                        <DfuThroughputGraph {...throughput} />
                    </div>
                </FormGroup>
            );
        }
        return null;
    }

    getStatus() {
        if (this.props.isCompleted) {
            return 'Completed';
        } else if (this.props.isStopping) {
            return 'Stopping...';
        }
        let status = this.props.status;
        const fileNameBeingTransferred = this.props.fileNameBeingTransferred;
        if (fileNameBeingTransferred) {
            status += ' ' + fileNameBeingTransferred;
        }
        return status + '...';
    }

    getPercentCompleted() {
        return this.props.isCompleted ? 100 : this.props.percentCompleted;
    }

    render() {
        const {
            onChooseFile,
            filePath,
            packageInfo,
            isStarted,
            isStopping,
            isCompleted,
            onStartDfu,
            onStopDfu,
        } = this.props;

        return (
            <form className='form-horizontal native-key-bindings'>
                <div className="col-md-12">
                    <FileInput
                        readOnly
                        buttonDisabled={isStarted || isCompleted}
                        label='Zip file'
                        value={filePath}
                        onChooseClicked={onChooseFile}
                    />
                    { packageInfo ? this.renderPackageInfo() : null }
                    { isStarted || isCompleted ? this.renderProgress() : null }
                    { isStarted || isCompleted ? this.renderGraph() : null }
                </div>
                <ButtonToolbar>
                    <div style={filePath && !isCompleted ? {} : {display: 'none'}}>
                        <DfuButton
                            dfuInProgress={isStarted}
                            disabled={isStopping}
                            onClick={isStarted ? onStopDfu : onStartDfu}/>
                    </div>
                </ButtonToolbar>
            </form>
        );
    }
}

DfuEditor.propTypes = {
    packageInfo: PropTypes.object,
    filePath: PropTypes.string,
    status: PropTypes.string,
    fileNameBeingTransferred: PropTypes.string,
    percentCompleted: PropTypes.number,
    isCompleted: PropTypes.bool,
    isStopping: PropTypes.bool,
    isStarted: PropTypes.bool,
    onChooseFile: PropTypes.func.isRequired,
    onStartDfu: PropTypes.func.isRequired,
    onStopDfu: PropTypes.func.isRequired,
    throughput: PropTypes.object,
};

export default DfuEditor;
