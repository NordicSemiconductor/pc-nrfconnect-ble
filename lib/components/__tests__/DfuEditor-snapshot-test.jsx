/* Copyright (c) 2016, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *   1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form, except as embedded into a Nordic
 *   Semiconductor ASA integrated circuit in a product or a software update for
 *   such product, must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 *
 *   3. Neither the name of Nordic Semiconductor ASA nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 *   4. This software, with or without modification, must only be used with a
 *   Nordic Semiconductor ASA integrated circuit.
 *
 *   5. Any software provided in binary form under this license must not be
 *   reverse engineered, decompiled, modified and/or disassembled.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
// Have to mock react-dom due to bug: https://github.com/facebook/react/issues/7386
// Can be removed after upgrade to react 15.4.0.
jest.mock('react-dom');

import React from 'react';
import renderer from 'react-test-renderer';
import DfuEditor from '../DfuEditor';

describe('DfuEditor', () => {
    it('should render with default props', () => {
        expect(renderComponent({})).toMatchSnapshot();
    });

    it('should render with file path', () => {
        expect(renderComponent({
            filePath: '/path/to/file'
        })).toMatchSnapshot();
    });

    it('should render with file path and package info', () => {
        expect(renderComponent({
            filePath: '/path/to/file',
            packageInfo: {
                application: {
                    'bin_file': 'dfu.bin',
                    'dat_file': 'dfu.dat',
                }
            }
        })).toMatchSnapshot();
    });

    it('should render with file path, package info, and dfu in progress', () => {
        expect(renderComponent({
            filePath: '/path/to/file',
            packageInfo: {
                application: {
                    'bin_file': 'dfu.bin',
                    'dat_file': 'dfu.dat',
                }
            },
            isStarted: true,
            percentCompleted: 10,
            status: 'In progress'
        })).toMatchSnapshot();
    });

    it('should render stopping', () => {
        expect(renderComponent({
            filePath: '/path/to/file',
            packageInfo: {
                application: {
                    'bin_file': 'dfu.bin',
                    'dat_file': 'dfu.dat',
                }
            },
            isStarted: true,
            isStopping: true,
            percentCompleted: 10
        })).toMatchSnapshot();
    });

    it('should render with dfu completed', () => {
        expect(renderComponent({
            filePath: '/path/to/file',
            packageInfo: {
                application: {
                    'bin_file': 'dfu.bin',
                    'dat_file': 'dfu.dat',
                }
            },
            isStarted: false,
            isCompleted: true
        })).toMatchSnapshot();
    });
});

function renderComponent(props) {
    return renderer.create(
        <DfuEditor
            isStarted={false}
            onChooseFile={() => {}}
            onStartDfu={() => {}}
            onStopDfu={() => {}}
            {...props}
        />
    ).toJSON();
}
