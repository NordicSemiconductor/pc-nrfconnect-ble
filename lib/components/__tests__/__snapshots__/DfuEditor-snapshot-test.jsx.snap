exports[`DfuEditor should render stopping 1`] = `
<form
  className="form-horizontal native-key-bindings">
  <div
    className="col-md-12">
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Zip file
      </label>
      <span
        className="col-md-9 input-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          readOnly={true}
          type={undefined}
          value="/path/to/file" />
        <span
          className="input-group-btn">
          <button
            className="btn btn-default"
            disabled={true}
            onClick={[Function]}
            type="button">
            Choose
          </button>
        </span>
      </span>
    </div>
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Package info
      </label>
      <span
        className="col-md-9 input-group">
        <pre>
          application:
            bin_file: dfu.bin
            dat_file: dfu.dat
          
        </pre>
      </span>
    </div>
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Progress
      </label>
      <span
        className="col-md-9 input-group">
        <p>
          Stopping...
        </p>
        <div
          className="progress">
          <div
            aria-valuemax={100}
            aria-valuemin={0}
            aria-valuenow={10}
            className="progress-bar"
            role="progressbar"
            style={
              Object {
                "width": "10%",
              }
            }>
            10%
          </div>
        </div>
      </span>
    </div>
  </div>
  <div
    className="btn-toolbar"
    role="toolbar">
    <div
      style={Object {}}>
      <button
        className="btn-nordic pull-right btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="button">
        <span
          className="icon-stop" />
        Stop DFU
      </button>
    </div>
  </div>
</form>
`;

exports[`DfuEditor should render with default props 1`] = `
<form
  className="form-horizontal native-key-bindings">
  <div
    className="col-md-12">
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Zip file
      </label>
      <span
        className="col-md-9 input-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          readOnly={true}
          type={undefined}
          value={undefined} />
        <span
          className="input-group-btn">
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button">
            Choose
          </button>
        </span>
      </span>
    </div>
  </div>
  <div
    className="btn-toolbar"
    role="toolbar">
    <div
      style={
        Object {
          "display": "none",
        }
      }>
      <button
        className="btn-nordic pull-right btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button">
        <span
          className="icon-play" />
        Start DFU
      </button>
    </div>
  </div>
</form>
`;

exports[`DfuEditor should render with dfu completed 1`] = `
<form
  className="form-horizontal native-key-bindings">
  <div
    className="col-md-12">
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Zip file
      </label>
      <span
        className="col-md-9 input-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          readOnly={true}
          type={undefined}
          value="/path/to/file" />
        <span
          className="input-group-btn">
          <button
            className="btn btn-default"
            disabled={true}
            onClick={[Function]}
            type="button">
            Choose
          </button>
        </span>
      </span>
    </div>
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Package info
      </label>
      <span
        className="col-md-9 input-group">
        <pre>
          application:
            bin_file: dfu.bin
            dat_file: dfu.dat
          
        </pre>
      </span>
    </div>
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Progress
      </label>
      <span
        className="col-md-9 input-group">
        <p>
          Completed
        </p>
        <div
          className="progress">
          <div
            aria-valuemax={100}
            aria-valuemin={0}
            aria-valuenow={100}
            className="progress-bar"
            role="progressbar"
            style={
              Object {
                "width": "100%",
              }
            }>
            100%
          </div>
        </div>
      </span>
    </div>
  </div>
  <div
    className="btn-toolbar"
    role="toolbar">
    <div
      style={
        Object {
          "display": "none",
        }
      }>
      <button
        className="btn-nordic pull-right btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button">
        <span
          className="icon-play" />
        Start DFU
      </button>
    </div>
  </div>
</form>
`;

exports[`DfuEditor should render with file path 1`] = `
<form
  className="form-horizontal native-key-bindings">
  <div
    className="col-md-12">
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Zip file
      </label>
      <span
        className="col-md-9 input-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          readOnly={true}
          type={undefined}
          value="/path/to/file" />
        <span
          className="input-group-btn">
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button">
            Choose
          </button>
        </span>
      </span>
    </div>
  </div>
  <div
    className="btn-toolbar"
    role="toolbar">
    <div
      style={Object {}}>
      <button
        className="btn-nordic pull-right btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button">
        <span
          className="icon-play" />
        Start DFU
      </button>
    </div>
  </div>
</form>
`;

exports[`DfuEditor should render with file path and package info 1`] = `
<form
  className="form-horizontal native-key-bindings">
  <div
    className="col-md-12">
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Zip file
      </label>
      <span
        className="col-md-9 input-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          readOnly={true}
          type={undefined}
          value="/path/to/file" />
        <span
          className="input-group-btn">
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button">
            Choose
          </button>
        </span>
      </span>
    </div>
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Package info
      </label>
      <span
        className="col-md-9 input-group">
        <pre>
          application:
            bin_file: dfu.bin
            dat_file: dfu.dat
          
        </pre>
      </span>
    </div>
  </div>
  <div
    className="btn-toolbar"
    role="toolbar">
    <div
      style={Object {}}>
      <button
        className="btn-nordic pull-right btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button">
        <span
          className="icon-play" />
        Start DFU
      </button>
    </div>
  </div>
</form>
`;

exports[`DfuEditor should render with file path, package info, and dfu in progress 1`] = `
<form
  className="form-horizontal native-key-bindings">
  <div
    className="col-md-12">
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Zip file
      </label>
      <span
        className="col-md-9 input-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          readOnly={true}
          type={undefined}
          value="/path/to/file" />
        <span
          className="input-group-btn">
          <button
            className="btn btn-default"
            disabled={true}
            onClick={[Function]}
            type="button">
            Choose
          </button>
        </span>
      </span>
    </div>
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Package info
      </label>
      <span
        className="col-md-9 input-group">
        <pre>
          application:
            bin_file: dfu.bin
            dat_file: dfu.dat
          
        </pre>
      </span>
    </div>
    <div
      className="form-group">
      <label
        className="col-md-3 text-right control-label"
        htmlFor={undefined}>
        Progress
      </label>
      <span
        className="col-md-9 input-group">
        <p>
          In progress...
        </p>
        <div
          className="progress">
          <div
            aria-valuemax={100}
            aria-valuemin={0}
            aria-valuenow={10}
            className="progress-bar"
            role="progressbar"
            style={
              Object {
                "width": "10%",
              }
            }>
            10%
          </div>
        </div>
      </span>
    </div>
  </div>
  <div
    className="btn-toolbar"
    role="toolbar">
    <div
      style={Object {}}>
      <button
        className="btn-nordic pull-right btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button">
        <span
          className="icon-stop" />
        Stop DFU
      </button>
    </div>
  </div>
</form>
`;
